### DISCUSSION TITLE
Use Typebox to define schema of a response which includes objects like a Date

### DISCUSSION THREAD

**Discussion by thhermansen** [fastify/fastify#3357] [Q&A] [answered]
Hi,

We have just started using Fastify in our application and we like it a lot. However, one hurdle we haven't figured out yet is how to express with Typebox a response which includes dates.

Lets take an example with a request handler simply returning an object containing the date of the request:

```ts
import {Static, Type} from '@sinclair/typebox'
import {FastifyPluginAsync} from 'fastify'

/**
 * Response schema, expressed via Typebox
 */
const Response = Type.Object({
  requestedOn: Type.String({format: 'date'}),
})

/**
 * Typescript type of the Reply
 */
type IResponse = Static<typeof Response>

const test: FastifyPluginAsync = async (fastify): Promise<void> => {
  fastify.get<{Reply: IResponse}>(
    '/test',
    {
      schema: {
        response: {
          '200': Response,
        },
      },
    },
    async () => {
      return {requestedOn: new Date()}
    }
  )
}

export default test
```

This works when ran; the response to a request is:

```bash
$ curl http://localhost/test
{"requestedOn":"2021-10-04"}
```

However, the request handler function:

```ts
    async () => {
      return {requestedOn: new Date()}
    }
```

yields a TS error:

```
          Types of property 'requestedOn' are incompatible.
            Type 'Date' is not assignable to type 'string'.
```

This makes sense as we have stated via Typebox that requestedOn should be a string and it's format is 'date'. However, the request handler function must return an object where `requestedOn` is a date in order to pass that on to `fast-json-string` so it can serialize the Date instance accordingly.

## So the question is 🥁🥁🥁

What am I missing here? How can we make Typescript happy in this scenario? :-)

**Comment by adamschoenemann** [comment]
I'd be very interested in this as well! :)

**Comment by RafaelGSS** [comment]
It's is more related to the `Typebox` type definition. @sinclairzx81 I don't know how to reproduce it locally, but looks like the `Static<Type.String({ format: 'date' })>` returns `string` instead of `string | Date`

**Comment by sinclairzx81** [comment]
@RafaelGSS Hi. TypeBox doesn't provide out of the box support for `Date` objects as dates generally require construction after deserialization. However if Fastify has infrastructure to automatically serialize / deserialize `Date` objects based on the `format`, you can create a custom type by extending `TSchema` in the following way.

TypeScript Example [Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQMoGMAWBTEBDAGjlRjxmHSOQE8xs4BfOAMyghDgHIABAZ2AB26ADZ5gUAPQxa2AEYQAHpwBQyiRLgBabTt179Bw0eO61GgCJlsNOma0mHjp4dXYFkWHEExsUZnnR6ZEsfG3o3HwEAE14UDBx8RGU4OAASXlJydAAuOBD6FJT1OBgcEpkSiC8BZl8AOmTyulyUzgyoQQBzTiKNACteCAE4XixcPCbsRuZofBhcziirHsLi9q6WWbJlBld3aHhmAFchciG8qzCACgBKXODLioRG4oEIH1z0zIo4MDY6WDUOBveBuP7YXi8PCyYT0QZwKDYGBHKDDPACaiNRHI1GISYLdYCbpEGZQOYLJY+HpMPCxdGY3Z2ZzMlkmOwASRqvmwQimxVZAsF2lU6CGGRQcAAvCgZHUAPKyPrYdAwK7PFLAKL3WUkDpE24ERqU7C5fLXG47C3KaR0CXSkhkCgAHht2AgzBQAD5VMVXXaksVCl4tSMYHrOgBuOxB42mqxR4oMIA)

```typescript
import { TSchema, Static, Type } from '@sinclair/typebox'

// -----------------------------------------------------------
// DateType
// -----------------------------------------------------------

export interface TDateType extends TSchema {
  $static: Date     // the type to infer.
  type:   'string'  // json schema type
  format: 'date'    // string format
}

export function DateType(): TDateType {
  // note: $static property not expressable so return any
  return { type: 'string', format: 'date' } as any
}

// -----------------------------------------------------------
// Inference
// -----------------------------------------------------------

const T = Type.Object({
  id: Type.String(),
  date: DateType()
})

type T = Static<typeof T>

// type T = {
//     id: string;
//     date: Date;
// }
```

Alternatively, you can extend the TypeBox TypeBuilder which will allow you to group application specific types under the `Type.*` namespace which can be more convenient than importing custom type functions. The following is an example of how to approach that.

TypeScript Example [Here](https://www.typescriptlang.org/play?#code/PTAEFpK6dv4Y6AoEoBiBDAzgFwJYBmAngCrEAOApgEICu+ANgCZUBOqYS3PvCyyfAFsKAeza5QAb1Dlq9JqzYAaWQGUAxgAsqQzKrW5MBDaAC+oQm1FDQAcgAC2fADsNjTPjbBclKgCNRAA87ASogsQlQACpLa1tHZzcPLx8-QJCwiPFJV1x2QkwNKlkAEWMS8PyXZmx1bV1MaWRQUAASPGN8DQAuUHL81qGW0F9qPta7PDZXAHNQ1sJxPVw+u2YK0LMsyNyXfLZC4tkAVTyADgBBNjZMYlAqqhq60k0dPWbWjqMTPrP9q43O5DVojMZUCb2UT+ABWVA0uAWoGw1A0+EwjHwAC8qMw1v9cIDbsQtjscqB3Dg6lg8EQyH4FCx2A8gtVarIGQwmWxPqAKHR-JjTAMqAAKACUfVIIt5rTYVFwdDYLmRdGobAAdIZxGKZOC1tM5nZVEs2Cs1ht8nZzOKRttWvzBd1QASiXcJVLXddibLQPLFcrVeqtbgdaK9X41tC4QjjcjUejMTi8fYvUCSTa7chtpw-VRwOFdshC+SNKIXHgOdRQABeUAuKgAdwwOAIJDktC5SgloAEaD4A8HSFzAEkXIR2E9irmh7O55ABGWK5JSLWq1QNQB5WHw3Dhkb4FMdkMzFyzCXKEaWiHrjUii8jfx0QgTthSvwatPEiXZ23IcGyGuhhdBoAA84KiIQsgAHx9mAAGrnWUi5iCh59IaZ4ANwoUM159CK2FoCCT4vuwfwXN6dy5mYQA)

```typescript
// -----------------------------------------------------------
// FastifyTypeBuilder
// -----------------------------------------------------------

import { TypeBuilder, TSchema, Static } from '@sinclair/typebox'

export * from '@sinclair/typebox'

export interface TDate extends TSchema {
  $static: Date     
  type:   'string'
  format: 'date'
}

export interface TUint8Array extends TSchema {
  $static: Uint8Array     
  type:   'object'
  specialized: 'Uint8Array'
}

export class FastifyTypeBuilder extends TypeBuilder {
  public Date(): TDate {
    return super.Store({ type: 'string', format: 'date' })
  }
  public Uint8Array(): TUint8Array {
    return super.Store({ type: 'object', specialized: 'Uint8Array' })
  }
}

// re-export
export const Type = new FastifyTypeBuilder() 

// -----------------------------------------------------------
// Inference
// -----------------------------------------------------------

const T = Type.Object({
  id: Type.String(),
  date: Type.Date(),
  buffer: Type.Uint8Array()
})

type T = Static<typeof T>

// type T = {
//     id: string;
//     date: Date;
//     buffer: Uint8Array
// }
```
Hope that helps!
S

**Reply by RafaelGSS** [reply]
Maybe I'm missing something, but, at least for me, it seems a Typescript fix.

The JSON Schema that `typebox` is generating is totally correct, it's working fine as said initially. @thhermansen is hitting a typescript barrier when `Static<Type.String({ format: 'date' })>` is returning a type `string` instead of `Date`. Nowadays, Fastify already supports that serialization.

**Reply by sinclairzx81** [reply]
@RafaelGSS TypeBox can't infer `Date` objects from `Type.String({ format: 'date' })`, it can only reasonably infer `string` types. The reason is the validation is testing specifically for an `ISO8601` string value and it's not possible for TypeBox to make assumptions about downstream applications having automatic `ISO8601` to `Date()` conversions during serialization. Such conversions are typically handled at a application or framework level (as is the case for Fastify)

For applications that do support implicit conversions, you can create a custom TypeBox type as above. The following extends the previous example above to support both `Date | string` unions, and wires it up using Type Providers.

```typescript
import { TSchema, Static, Type } from '@sinclair/typebox'

// -----------------------------------------------------------
// DateType
// -----------------------------------------------------------

export interface TDateType extends TSchema {
  $static: Date | string // allow both Date and string
  type:   'string' 
  format: 'date'
}

export function DateType(): TDateType {
  return { type: 'string', format: 'date' } as any
}

// -----------------------------------------------------------
// Route
// -----------------------------------------------------------

fastify.get('/date', {
  schema: {
      response: {
         200: DateType()
      }
  }
}, (req, res) => {
  res.send(new Date())    // OK: Allow Date objects to be sent (implicit conversion to ISO8601)
  
  res.send('2020-04-04')  // OK: Allow ISO8601
})
```
Given that Fastify does support `string` to `Date` conversions, I'm wondering if it might be reasonable to create a `FastifyTypeBuilder` (as per previous example) that implements a `Type.Date()` type and exports it on the `fastify-typebox-type-provider` package.

**Reply by RafaelGSS** [reply]
Thanks for the clarification.

However, I'm afraid to add that dependency in `fastify-typebox-type-provider`. Couldn't we get a `Type.ISO8601()` in `typebox`?

**Comment by sinclairzx81** [comment]
Response To: https://github.com/fastify/fastify/discussions/3357#discussioncomment-2442328

@RafaelGSS Unfortunately I can't. TypeBox can only provide type inference for data types that can be naturally expressed in JSON (and Date objects fall outside this category). It really cannot make assumptions about how downstream applications will choose to serialize / deserialize data transmitted through JSON, be it Date, Uint8Array, Mongo ObjectId, etc. ..... Rather it provides the above extension mechanisms that enable users to create their own custom types and define custom inference rules for those types (with the custom inference rules tied to custom deserialization implemented by the application)

> Note: Providing simpler mechanisms for users to create and extend types is an ongoing work-in progress. Some of which is currently being explored in [this](https://github.com/sinclairzx81/sidewinder) project.

As an aside, wouldn't converting to `toISOString()` be the simplest solution to this particular query?

```typescript
const test: FastifyPluginAsync = async (fastify): Promise<void> => {
  fastify.get<{Reply: IResponse}>(
    '/test',
    {
      schema: {
        response: {
          '200': Response,
        },
      },
    },
    async () => {
      return {requestedOn: new Date().toISOString() } // convert to iso string ?
    }
  )
}
```

**Reply by RafaelGSS** [reply]
Yes, I just summarized it in a new reply. Thank you again @sinclairzx81!

**Comment by RafaelGSS** [comment]
I think @sinclairzx81  provided two great suggestions, just to summarize here:

1) 

```ts
import { TSchema, Static, Type } from '@sinclair/typebox'

// -----------------------------------------------------------
// DateType
// -----------------------------------------------------------

export interface TDateType extends TSchema {
  $static: Date | string // allow both Date and string
  type:   'string' 
  format: 'date'
}

export function DateType(): TDateType {
  return { type: 'string', format: 'date' } as any
}

// -----------------------------------------------------------
// Route
// -----------------------------------------------------------

fastify.get('/date', {
  schema: {
      response: {
         200: DateType()
      }
  }
}, (req, res) => {
  res.send(new Date())    // OK: Allow Date objects to be sent (implicit conversion to ISO8601)
  
  res.send('2020-04-04')  // OK: Allow ISO8601
})
```

2) 
```ts
const test: FastifyPluginAsync = async (fastify): Promise<void> => {
  fastify.get<{Reply: IResponse}>(
    '/test',
    {
      schema: {
        response: {
          '200': Response,
        },
      },
    },
    async () => {
      return {requestedOn: new Date().toISOString() } // convert to iso string ?
    }
  )
}
```

**Comment by sinclairzx81** [comment, marked as answer]
Just a quick update on this. As of `0.24.x`, TypeBox now provides a `Type.Unsafe()` function that can be used to create custom schemas with custom inference rules. From this, `Date` objects can be safely inferred if the underlying framework is known to automatically convert deserialized values....

The following shows it's usage.

```typescript
const T = Type.Object({
   created: Type.Unsafe<Date>({ type: 'string', format: 'date' }) // created inferred as Date
})
```

Additional information on this feature can be found [here](https://github.com/sinclairzx81/typebox#unsafe-types)

**Reply by james-gardner** [reply]
As of 0.25.x there's `Type.Date`. See: https://github.com/sinclairzx81/typebox/blob/master/changelog.md

**Reply by sinclairzx81** [reply]
@james-gardner Hi, Good spotting! :) Yes, I did actually cave (temporarily) and provide the non-standard `Type.Date()` and `Type.Uint8Array()` types due to these types being routinely requested for other IO / In-Memory use cases (situations where serialization of data was not a factor)

These types are intended to be used in the following scenarios.

- You're using the TypeBox TypeCompiler
- You're using Ajv with custom configurations
- You have Date and UInt8Array message serialization support (using MsgPack instead of JSON)
- You need to validate in-memory Date objects prior to writing to a Database (such as MongoDB)

In Fastify, the recommendation is to continue to use either `Type.Unsafe<Date>({ type: 'string', format: 'date' }) ` as per suggestion, or encoding Date's as numbers (timestamp) or strings (iso8601) via `Type.String({ format: 'date-time' })`. 

### User Defined Types

Just as a historical note, when the Date type was added to TypeBox, there was actually a [larger discussion](https://github.com/sinclairzx81/typebox/discussions/263) happening around being able to support User defined types in TypeBox (of which Date would be an example). At the time of implementing Date, TypeBox did not support augmenting the compiler and value checking API's (this functionality wasn't planned for until next year), but has made it in early due to some other requirements to be able to support types above and beyond dates and buffers.

As of `typebox@0.25.10`, the following is possible. [Documentation Here](https://github.com/sinclairzx81/typebox#typecheck-custom-types)

```typescript
Custom.Set('Date', (schema, value) => value instanceof Date) // custom type check

const T = Type.Unsafe<Date>({ [Kind]: 'Date' })              // custom type schema

const R = Value.Check(T, new Date())                         // true
```
Because of this, I am going to be looking to deprecate both `Type.Date()` and `Type.UInt8Array()` and recommend these types be expressed as user defined types in future.

Hope that brings a bit of insight into the inclusion of `Type.Date` (and why you probably shouldn't use it) :D 
Cheers
S

**Reply by ghost** [reply]
Without registering any TypeBox formats with `FormatRegistry` in a Fastify project from last year, I had to use `datetime` instead of `date-time` as the format string and I'm still not sure why. Here's what I came up with and if I figure out what the cause is, I'll post it here for future readers.

```ts

/**
 * Defines a JS `Date` field that is safe to use with Sequelize, TypeScript and
 * as an OpenAPI schema.
 *
 * Serializes `Date`s as ISO-8601 (with time). Pass `{ format: "date" }` to
 * exclude the time.
 *
 * See *Fastify discussion*
 * [Use Typebox to define schema of... Date](https://github.com/fastify/fastify/discussions/3357#discussioncomment-4241449):
 *
 * > In Fastify, the recommendation is to continue to use either
 * > `Type.Unsafe<Date>({ type: 'string', format: 'date' })` as per suggestion,
 * > or encoding Date's as numbers (timestamp) or strings (iso8601) via
 * > `Type.String({ format: 'date-time' })`.
 *
 * @param options Options for `Type.Unsafe()`.
 * @returns A safe date type schema.
 *
 */
export function SafeDate(
  options?: Parameters<typeof Type.Unsafe<Date>>[0] & {
    /** The date serialization format. Defaults to `datetime` (ISO-8601). */
    format: "date" | "datetime";
  },
) {
  return Type.Unsafe<Date>({ type: "string", format: "datetime", ...options });
}
```

For my new project, I'm going to try registering all of the formats from Ajv [detailed here](https://github.com/fastify/fastify-type-provider-typebox/issues/84#issuecomment-1560337451) and to use `setValidatorCompiler(TypeBoxValidatorCompiler)` ( from fastify/fastify-type-provider-typebox ) on my Fastify instance which I had not done before.
